### DISCUSSION TITLE
Chunkservers not rebalancing (ec4+1)

### DISCUSSION THREAD

**Discussion by Lindsay-Mathieson** [moosefs/moosefs#600] [Q&A]
Had my system on Goal 2 and all 6 chunkservers were perfectly balanced at 43%

Changed everything to ec 4+1, that went well, all chunks converted over a couple of days :) But 4 chunkservers are at 35.4% and 2 are at 20.9%, with no writes happening

Could it because of the uneven disk size distribution in my chunkservers?

<img width="707" alt="image" src="https://github.com/user-attachments/assets/ee410e5c-6921-4e69-b66b-d17a31b7dc9c">

<img width="706" alt="image" src="https://github.com/user-attachments/assets/2bd45e08-5b2a-4472-8f3a-4862b03178d4">

<img width="662" alt="image" src="https://github.com/user-attachments/assets/7a8bf887-7cff-4e39-bbc9-f4829f52d81c">

**Comment by tianon** [comment]
I bet your system has 211,353 unique chunks total (the number on those two bigger drives) so they've each got a copy of every chunk and thus can't possibly be any fuller. 😅

**Reply by tianon** [reply]
Oh sorry, yep that's confirmed by your last screenshot. It can't balance any further because there's nothing more that can move to your bigger drives without them having *two* copies of a chunk, which would be kind of silly. 😄

**Reply by Lindsay-Mathieson** [reply]
Yup!

Does that mean I'm constrained by the max chunks that can be placed on the first four chunkservers? My effective usage is 35% ?

I guess my alternatives are:

* Add another chunkserver (16TB?)
* or
* Bring the first four chunkservers up to 16TB each (15TB effective)

**Reply by chogata** [reply]
Your MooseFS has put the max possible number of chunk copies/parts on the 2 biggest chunk servers and cannot put anything more there, without breaking the "only one part or copy of chunk per server" rule. And it can't break this rule, because this rule means your files are safe whenever one chunk server goes offline :) But the max possible number of chunks is still lower percentage of capacity, because the capacity of your 6 chunk servers differs so much. Adding another 16TB chunk server will mean stuff will get copied there, but still, new chunks will be recorded on all 6 chunk servers, so the balance will not last. The only way to keep the occupied space balanced is to have more or less the same capacity in each chunk server, especially, if you don't have many of them. With 12+ chunk server MooseFS would be better able to balance servers with varying capacity.

**Reply by Lindsay-Mathieson** [reply]
> Your MooseFS has put the max possible number of chunk copies/parts on the 2 biggest chunk servers and cannot put anything more there, without breaking the "only one part or copy of chunk per server" rule.

Thanks, my best option seems to be to upgrade the first four chunkservers, very tempting 😁and a great excuse.

Reality is I could keep going for a while as is 😢

**Comment by Lindsay-Mathieson** [comment]
Ok, I brought all 6 chunkservers up to 16TB - as part of the shuffle I converted back to Goal 2, because I lacked drive bays to add all the disks, had to move most of the chunks to a few chunkservers, add the new disks, then rebalance - took a while :)

Now on 6*16TB, Goal 2, 28% usage.

TBH, I could keep going for a while like this. Tempted to switch back to ec4+1, but it seems that leaves me more vulnerable. If I lose one disk then I'm down to 5 chunkservers and the system is vulnerable, it can't rebalance, whereas if I stayed with Goal 2, then the loss of a chunkserver just rebalances all the data.

Am I misreading this?

**Reply by chogata** [reply]
You are misreading it ;)

6 chunk servers, data in 2 copies -> one chunk server goes offline (5 CS left), some data is now in 1 copy -> missing copies are eventually duplicated from existing ones and everything is back in 2 copies, chunks can be rebalanced between servers

6 chunk servers, data partially in 2 copies, partially in EC4+1 format, conversion between copy format and EC+1 format works freely back and forth -> one chunk server goes offline (5 CS left), some data is now in 1 copy or is missing 1 part -> missing copies are eventually duplicated from existing ones, missing parts are eventually restored using the existing ones, conversion only works now in one direction, that is from EC+1 to 2 copies (if needed for I/O), but there is no conversion from copies to EC, chunks and parts can be rebalanced between servers

So everything works, everything has the same level of security in both situations, the only thing is that you loose the ability to convert any "new" (not new as in newly recorded, but new as in not yet converted) chunks to EC format. Once the sixth chunk server is back on-line, the conversion will again become 2-way.

OTOH if your data now uses only 28% of your instance's capacity, you don't have to use EC. EC is nice to save space, but copies work as well (even better, there were a lot of optimisations!) as they used to, so if you don't need to save space, you can stay with copies.

**TL;DR:** your system is as safe with 2 copies as with 2 copies mixed with EC4+1 when it comes to one chunk server failure scenario.

**Reply by Lindsay-Mathieson** [reply]
> OTOH if your data now uses only 28% of your instance's capacity, you don't have to use EC. EC is nice to save space, but copies work as well (even better, there were a lot of optimisations!) as they used to, so if you don't need to save space, you can stay with copies.
> 
> **TL;DR:** your system is as safe with 2 copies as with 2 copies mixed with EC4+1 when it comes to one chunk server failure scenario.

Thank you for the detailed explanation! MooseFS has had a lot of thought and design put into it, really impressive how it falls back to data safety.

I'll stick with 2 copies for now as I have a lot of breathing room. This may change as I upgrade my content to 4K though 😁
